### PR TITLE
Make udev link re-renaming conditional

### PR DESCRIPTION
**Pull request by keszybz**

### CODE REVIEW

**Comment by evverx** [comment]
This pull request **introduces 1 alert** when merging 75b13196120116783394db2b3ae2ccc521401779 into 35f2c0ba6afd70c4a1f74865d9231b85f4a01380 - [view on LGTM.com](https://lgtm.com/projects/g/systemd/systemd/rev/pr-92b3a6e3582b5d06e522acf49a08c9f3f9ff52e6)

**new alerts:**

* 1 for Missing header guard

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by keszybz** [comment]
Updated to add missing pragma.

File: src/udev/net/link-config.c [hunk 1]
```diff
@@ -352,12 +333,13 @@ static int get_mac(sd_device *device, bool want_random,
...
 int link_config_apply(link_config_ctx *ctx, link_config *config,
                       sd_device *device, const char **name) {
-        bool respect_predictable = false;
         struct ether_addr generated_mac;
         struct ether_addr *mac = NULL;
+        NamePolicy policy;
         const char *new_name = NULL;
         const char *old_name;
         unsigned speed;
+        unsigned name_type = NET_NAME_UNKNOWN;
```
Comment (poettering) [inline src/udev/net/link-config.c]: merge with previous line?

File: src/udev/net/link-config.c [hunk 2]
```diff
@@ -410,42 +392,56 @@ int link_config_apply(link_config_ctx *ctx, link_config *config,
...
-                                        break;
-                                case NAMEPOLICY_PATH:
-                                        (void) sd_device_get_property_value(device, "ID_NET_NAME_PATH", &new_name);
-                                        break;
-                                case NAMEPOLICY_MAC:
-                                        (void) sd_device_get_property_value(device, "ID_NET_NAME_MAC", &new_name);
-                                        break;
-                                default:
-                                        break;
+        (void) link_name_type(device, &name_type);
```
Comment (poettering) [inline src/udev/net/link-config.c:401]: i think adding debug logging here if this fails would make a ton of sense

File: src/udev/net/link-config.c [hunk 3]
```diff
@@ -410,42 +392,56 @@ int link_config_apply(link_config_ctx *ctx, link_config *config,
...
                         }
                 }
-        }
 
-        if (should_rename(device, respect_predictable)) {
-                /* if not set by policy, fall back manually set name */
-                if (!new_name)
-                        new_name = config->name;
+        if (new_name)
+                log_device_info(device, "Policy *%s* yields \"%s\".", name_policy_to_string(policy), new_name);
```
Comment (poettering) [inline src/udev/net/link-config.c]: maybe LOG_DEBUG?

File: src/udev/net/naming-scheme.c [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+static const NamingScheme naming_schemes[] = {
+        { "v238", NAMING_V238 },
+        { "v239", NAMING_V239 },
+        { "v240", NAMING_V240 },
+        /* … add more schemes here, as the logic to name devices is updated … */
+};
+
+
+static const NamingScheme* naming_scheme_from_name(const char *name) {
```
Comment (poettering) [inline src/udev/net/naming-scheme.c:14]: spurious double newline

**Comment by poettering** [comment]
i think a small NEWS additoinal about this would be good.

File: src/udev/net/link-config.c [hunk 5]
```diff
@@ -410,42 +392,56 @@ int link_config_apply(link_config_ctx *ctx, link_config *config,
...
+                for (NamePolicy *p = config->name_policy; !new_name && *p != _NAMEPOLICY_INVALID; p++) {
+                        policy = *p;
+
+                        switch (policy) {
+                        case NAMEPOLICY_KERNEL:
+                                if (name_type != NET_NAME_PREDICTABLE)
+                                        continue;
+
+                                /* The kernel claims to have given a predictable name, keep it. */
+                                log_device_info(device, "Policy *%s*: keeping predictable kernel name",
```
Comment (yuwata) [inline src/udev/net/link-config.c]: I think `LOG_DEBUG` is enough for this.

File: src/udev/net/link-config.c [hunk 6]
```diff
@@ -414,6 +409,13 @@ int link_config_apply(link_config_ctx *ctx, link_config *config,
                                 log_device_info(device, "Policy *%s*: keeping predictable kernel name",
                                                 name_policy_to_string(policy));
                                 goto no_rename;
+                        case NAMEPOLICY_KEEP:
+                                if (!IN_SET(name_type, NET_NAME_USER, NET_NAME_RENAMED))
+                                        continue;
+
+                                log_device_info(device, "Policy *%s*: keeping existing userspace name",
```
Comment (yuwata) [inline src/udev/net/link-config.c]: Also this one.

File: src/udev/net/link-config.c [hunk 7]
```diff
@@ -396,6 +397,12 @@ int link_config_apply(link_config_ctx *ctx, link_config *config,
         (void) link_name_type(device, &name_type);
         log_device_debug(device, "Device has name_assign_type=%d", name_type);
 
+        if (IN_SET(name_type, NET_NAME_USER, NET_NAME_RENAMED)
+            && !naming_scheme_has(NAMING_ALLOW_RERENAMES)) {
+                log_device_info(device, "Device already has a name given by userspace, not renaming.");
```
Comment (yuwata) [inline src/udev/net/link-config.c]: Also this one.

**Comment by keszybz** [comment]
All updated, PTAL.

**Comment by yuwata** [comment]
Still two `log_device_info()`. That is not strong opinion, but why?

**Comment by yuwata** [comment]
Please update the following:
https://github.com/systemd/systemd/blob/850115b3a1119c87f14ecb7634615632ef0b1933/man/systemd.link.xml#L261-L262

File: src/udev/net/link-config.c (hunk 2 above)
Comment (poettering) [reply src/udev/net/link-config.c:401]: this has not been addressed?

Comment (keszybz) [reply src/udev/net/link-config.c:401]: It has. `link_name_type` logs on it's own. (That makes more sense because there's two failure points.)

Comment (poettering) [reply src/udev/net/link-config.c:401]: a true. ignore me.

**Comment by keszybz** [comment]
Updated. Sorry about the missed log levels: I updated on of the patches and forgot that the other one needs updating too. I changed the two remaining log_info's to log_debug as suggested by @yuwata. There doesn't seem to be a good reason to make those two different than the surrounding ones.

I also added a formatting patch on top make the man page nicer.

**Comment by martinpitt** [comment]
This brought back the naming regression: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=917128#72 . This is using systemd .netlink files with `Name=` assignments.

**Comment by yuwata** [comment]
@martinpitt Does 03dd7b7ddec1b0e06f254972a2e05f516a05edaf (#11729) fix the issue? BTW, please file the issue as a new issue.

**Pull request merged**